keps/sig-node/4742-node-topology-downward-api/README.md

```diff
@@ -0,0 +1,416 @@
+it can also be used to describe other types of information about the cluster.
+This KEP proposes to allow the expansion of the concept of topology to include
+user-defined aspects about their cluster nodes, and in turn provide a way for
+pods to receive this information.
+
+Workarounds today typically involve using an initContainer to query the
+Kubernetes API and then pass data via shared volume to other containers within
+the same pod. By comparison, this proposal would reduce the number of service
+accounts and API server clients. Another workaround is to use webhooks to inject
+labels into pods, but this relies on advance knowledge of where the pod is going
+to be scheduled and requires the webhook to be running and available at the time
+of pod creation. This proposal would provide an easier way to access node labels
+from pods, and would be more efficient than the current workarounds.
+
+### Goals
+
+* Values from labels `topology.k8s.io/*` and `kubernetes.io/hostname` are made
+  available via downward API 
+
+### Non-Goals
+
+* Not to expose additional node info outside of topology labels
+* Not to pass any additional node labels other than `topology.k8s.io/*` and
+  `*.topology.k8s.io/*` to pods
+* Not to update pod labels after the initial node -> pod copy has been made
+* Not to make assurances regarding timing and availability of the label beyond
+  the initial pod label copy at scheduling time
+* Not to make assurances about the immutability of the pod label after the
+  initial copy. As with other labels, the pod label can be updated by the user
+  after the pod is created.
+
+## Proposal
+
+KEP 1659 defines the following labels: `topology.kubernetes.io/region` and
+`topology.kubernetes.io/zone` to be used for topology information. These labels
+are useful for pods as well to be able to make application decisions based on
+the region or zone the pod is running in. This KEP proposes to make these labels
+available to pods while also expanding upon KEP 1659 to allow for user-defined
+labels in the `*.topology.kubernetes.io` namespace.
+
+This KEP expands on KEP 1659 in the following ways:
+
+1. Label prefixes of the form `<domain>.topology.kubernetes.io` are allocated
+   for use by end users. The Kubernetes project itself will not define any
+   labels with this prefix.
+2. Labels of the form `<domain>.topology.kubernetes.io/<field>` will be passed
+   to pods.
+3. Labels of the form `topology.kubernetes.io/*` will be passed to pods but will
+   continue to be reserved by the Kubernetes project.
+4. All labels with `topology.kubernetes.io` and `*.topology.kubernetes.io`
+   prefixes should be considered safe for pods and should only contain
+   information that pods and containers can safely consume.
+
+The idea is that we will expose those labels from nodes to pods via a literal
+copy from the node. From that point, the topology labels can be used in the same
```

Comment:
...from the node when the pod is scheduled?  Or when the pod is started by kubelet?

If scheduling: Does it become a responsibility of every scheduler to do this, or is it automatic?

If Kubelet: Does it get locked at pod admission-time or at pod (re)start time (i.e. could change if the pod restarts)?